src/main/scala/com/blackmorse/spark/clickhouse/datasourcev2/ClickhousePartitionReader.scala

```diff
@@ -0,0 +1,71 @@
+package com.blackmorse.spark.clickhouse.datasourcev2
+
+import com.clickhouse.jdbc.ClickHouseDriver
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow
+import org.apache.spark.sql.connector.read.{InputPartition, PartitionReader, PartitionReaderFactory}
+
+import java.sql.ResultSet
+import java.util.Properties
+
+class ClickhousePartitionReaderFactory(clickhouseReaderInfo: ClickhouseReaderInfo) extends PartitionReaderFactory {
+  override def createReader(partition: InputPartition): PartitionReader[InternalRow] =
+    new ClickhousePartitionReader(
+      clickhouseReaderInfo = clickhouseReaderInfo,
+      clickhouseInputPartition = partition.asInstanceOf[ClickhouseInputPartition])
+}
+
+class ClickhousePartitionReader(clickhouseReaderInfo: ClickhouseReaderInfo, clickhouseInputPartition: ClickhouseInputPartition) extends PartitionReader[InternalRow] with Logging {
+  private val fields = clickhouseReaderInfo.schema.fields.map(f => s"`${f.name}`").mkString(", ")
+  val sql = s"SELECT $fields FROM ${clickhouseReaderInfo.tableName}"
+
+  private val conn = new ClickHouseDriver().connect(clickhouseReaderInfo.url, new Properties())
+  private val stmt = conn.prepareStatement(sql, ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY)
+  stmt.setFetchSize(300)
```

Comment:
+ `.setQueryTimeout(seconds);`

Reply:
Will do in the task dedicated to timeout's tuning